Forge/src/main/java/me/ferdz/placeableitems/client/model/FluidModel.java

```diff
@@ -0,0 +1,133 @@
+     * present, it will be calculated and cached. Additionally, the model will be recalculated if the
+     * result of {@link #shouldRecalculate(BlockState)} is true.
+     *
+     * @param state the state whose model to retrieve
+     *
+     * @return the rendering bounds
+     */
+    public final AllVertexBoundingBox getRenderBounds(BlockState state) {
+        return shouldRecalculate(state) ? renderCache.compute(state, (s, existing) -> calculateRenderBounds(s)) : renderCache.computeIfAbsent(state, this::calculateRenderBounds);
+    }
+
+    /**
+     * Calculate the rendering bounds for this model. This method does not account for cached models
+     * and should be used sparingly. Method callers should use {@link #getRenderBounds(BlockState)}
+     * instead to ensure a cached value is retrieved instead.
+     * <p>
+     * Implementations should consider the state's rotation and other states the block may have. Each
+     * unique state will be cached in the model's render cache. This method will only be called such that
+     * no model has yet been cached for the provided state, or {@link #shouldRecalculate(BlockState)}
+     * returns true.
+     * <p>
+     * These bounds support negative values and will be relative to {@link #getRotationPoint(BlockState)}.
+     *
+     * @param state the state whose render bounds to calculate
+     *
+     * @return the rendering bounds
+     */
+    public abstract AllVertexBoundingBox calculateRenderBounds(BlockState state);
+
+    /**
+     * Check whether or not the provided state has been cached.
+     *
+     * @param state the state to check
+     *
+     * @return true if cached, false otherwise
+     */
+    public boolean isCached(BlockState state) {
+        return renderCache.containsKey(state);
+    }
+
+    /**
+     * Clear the render cache for this model. Models for each unique state will be lazily recalculated when
+     * required.
+     */
+    public void clearRenderCache() {
+        this.renderCache.clear();
+    }
+
+    /**
+     * Check whether or not a quad should be rendered according to the specified direction
+     *
+     * @param state the state to check
+     * @param direction the direction to check
+     *
+     * @return true if the direction should be rendered, false otherwise
+     */
+    public boolean shouldRender(BlockState state, Direction direction) {
+        return true;
+    }
+
+    /**
+     * Get a pixel-scaled vector (values ranging from 0 - 16) for which this model may be rotated.
+     * {@link #calculateRenderBounds(BlockState)} will be rendered relative to this point. This value defaults
+     * to (0, 0, 0), the pixel relative to the origin of the block
+     *
+     * @param state the state whose rotation point to get
+     *
+     * @return the rotation point
+     */
+    public Vector3d getRotationPoint(BlockState state) {
+        return ZEROED_OFFSET;
```

Comment:
I think this should be abstract, given that it makes no sense having a default (especially at zero)